RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,631 @@
+#include "RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h" 
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+MahiFit::MahiFit() :
+  fullTSSize_(19), 
+  fullTSofInterest_(8)
+{}
+
+double MahiFit::getSiPMDarkCurrent(double darkCurrent, double fcByPE, double lambda) const {
+  double mu = darkCurrent * 25 / fcByPE;
+  return sqrt(mu/pow(1-lambda,3)) * fcByPE;
+}
+
+void MahiFit::setParameters(bool iDynamicPed, double iTS4Thresh, double chiSqSwitch, 
+			    bool iApplyTimeSlew, HcalTimeSlew::BiasSetting slewFlavor,
+			    double iMeanTime, double iTimeSigmaHPD, double iTimeSigmaSiPM, 
+			    const std::vector <int> &iActiveBXs, int iNMaxItersMin, int iNMaxItersNNLS,
+			    double iDeltaChiSqThresh, double iNnlsThresh) {
+
+  dynamicPed_    = iDynamicPed;
+
+  ts4Thresh_     = iTS4Thresh;
+  chiSqSwitch_   = chiSqSwitch;
+
+  applyTimeSlew_ = iApplyTimeSlew;
+  slewFlavor_    = slewFlavor;
+
+  meanTime_      = iMeanTime;
+  timeSigmaHPD_  = iTimeSigmaHPD;
+  timeSigmaSiPM_ = iTimeSigmaSiPM;
+
+  activeBXs_ = iActiveBXs;
+
+  nMaxItersMin_  = iNMaxItersMin;
+  nMaxItersNNLS_ = iNMaxItersNNLS;
+
+  deltaChiSqThresh_ = iDeltaChiSqThresh;
+  nnlsThresh_    = iNnlsThresh;
+
+  bxOffsetConf_ = -(*std::min_element(activeBXs_.begin(), activeBXs_.end()));
+  bxSizeConf_   = activeBXs_.size();
+
+}
+
+void MahiFit::phase1Apply(const HBHEChannelInfo& channelData,
+			  float& reconstructedEnergy,
+			  float& reconstructedTime,
+			  bool& useTriple, 
+			  float& chi2) const {
+
+  assert(channelData.nSamples()==8||channelData.nSamples()==10);
+
+  resetWorkspace();
+
+  nnlsWork_.tsSize = channelData.nSamples();
+  nnlsWork_.tsOffset = channelData.soi();
+  nnlsWork_.fullTSOffset = fullTSofInterest_ - nnlsWork_.tsOffset;
+
+  // 1 sigma time constraint
+  if (channelData.hasTimeInfo()) nnlsWork_.dt=timeSigmaSiPM_;
+  else nnlsWork_.dt=timeSigmaHPD_;
+
+  //Dark current value for this channel (SiPM only)
+  float darkCurrent =  getSiPMDarkCurrent(channelData.darkCurrent(), 
+					  channelData.fcByPE(),
+					  channelData.lambda());
+
+  //Average pedestal width (for covariance matrix constraint)
+  float pedVal = 0.25*( channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)+
+			channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)+
+			channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)+
+			channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3) );
+
+  nnlsWork_.pedConstraint = pedVal*SampleMatrix::Ones(nnlsWork_.tsSize, nnlsWork_.tsSize);
+  nnlsWork_.amplitudes.resize(nnlsWork_.tsSize);
+  nnlsWork_.noiseTerms.resize(nnlsWork_.tsSize);
+
+  std::array<float,3> reconstructedVals;
```

Comment:
please initialize to sensible values here.
It looks like there is no promise in the fit itself to fill them.